randxdr/marshaller.go

```diff
@@ -0,0 +1,121 @@
+	presets      []Preset
+}
+
+func (*randMarshaller) Sprintf(f string, args ...interface{}) string {
+	return fmt.Sprintf(f, args...)
+}
+
+func (rm *randMarshaller) randomKey(m interface{}) int32 {
+	keys := reflect.ValueOf(m).MapKeys()
+	return int32(keys[rm.rand.Intn(len(keys))].Int())
+}
+
+func (rm *randMarshaller) applyPreset(field string, i goxdr.XdrType) bool {
+	for _, preset := range rm.presets {
+		if preset.Selector(field, i) {
+			switch goxdr.XdrBaseType(i).(type) {
+			case goxdr.XdrEnum, goxdr.XdrNum32:
+				rm.useTag = false
+			}
+
+			preset.Setter(rm, field, i)
+			return true
+		}
+	}
+	return false
+}
+
+// Marshal populates a given goxdr.XdrType with random values.
+//
+// Every complex goxdr.XdrType has functions like XdrRecurse() which
+// allow you to visit subfields of the complex type. That is how
+// randMarshaller is able to populate subfields of a complex type with
+// random values.
+//
+// Note that randMarshaller is stateful because of how union types are handled.
+// Therefore Marshal() should not be used concurrently.
+func (rm *randMarshaller) Marshal(field string, i goxdr.XdrType) {
+	if rm.applyPreset(field, i) {
+		return
+	}
+
+	switch t := goxdr.XdrBaseType(i).(type) {
+	case goxdr.XdrVarBytes:
+		bound := t.XdrBound()
+		if bound > rm.maxBytesSize {
+			bound = rm.maxBytesSize
+		}
+		bound++
+		bs := make([]byte, rm.rand.Uint32()%bound)
+		rm.rand.Read(bs)
+		t.SetByteSlice(bs)
+	case goxdr.XdrBytes:
+		// t.GetByteSlice() returns the underlying byte slice
+		// rm.rand.Read() will fill that byte slice with random values
+		rm.rand.Read(t.GetByteSlice())
+	case goxdr.XdrVec:
+		bound := t.XdrBound()
+		if bound > rm.maxVecLen {
+			bound = rm.maxVecLen
+		}
+		bound++
+		vecLen := rm.rand.Uint32() % bound
+		t.SetVecLen(vecLen)
+		t.XdrMarshalN(rm, field, vecLen)
+	case goxdr.XdrPtr:
+		present := rm.rand.Uint32()&1 == 1
+		t.SetPresent(present)
+		t.XdrMarshalValue(rm, field)
+	case *goxdr.XdrBool:
+		t.SetU32(rm.rand.Uint32() & 1)
+	case goxdr.XdrEnum:
+		if rm.useTag {
+			rm.useTag = false
+			t.SetU32(rm.tag)
+		} else {
+			t.SetU32(uint32(rm.randomKey(t.XdrEnumNames())))
+		}
+	case goxdr.XdrNum32:
+		if rm.useTag {
+			rm.useTag = false
+			t.SetU32(rm.tag)
+		} else {
+			t.SetU32(rm.rand.Uint32())
+		}
+	case goxdr.XdrNum64:
+		t.SetU64(rm.rand.Uint64())
+	case goxdr.XdrUnion:
+		// If we have an XDR union we need to set the tag of the union.
+		// However, there is no SetTag() function in the goxdr.XdrUnion interface.
+		// We must rely on these two facts:
+		// * when XdrRecurse() is called on a union, the first field which will be marshalled is the tag field
+		// * the tag field can be one of two types: uint32 or enum
+		if m := t.XdrValidTags(); m != nil {
+			rm.tag = uint32(rm.randomKey(m))
+			rm.useTag = true
+			// The next field the marshaller will visit is the tag field.
+			// Once the tag is set, we need to toggle rm.useTag to false.
+		}
+		t.XdrRecurse(rm, field)
```

Comment:
Nit: You could potentially avoid the statefulness and make this thread-safe by passing a new `randomMarshaller` to this function. Something like:
```suggestion
		t.XdrRecurse(&randomMarshaller{
			tag: uint32(rm.randomKey(m)),
			rand: rm.rand,
			maxBytesSize: rm.maxBytesSize,
			maxVecLen: rm.maxVecLen,
			presets: rm.presets,
		}, field)
```

Then you don't need the `useTag` thing at all.

Reply:
I don't think that would that would work. Here is what  `XdrRecurse()` looks like for a union:

```golang
func (u *LedgerCloseMeta) XdrRecurse(x XDR, name string) {
	if name != "" {
		name = x.Sprintf("%s.", name)
	}
	XDR_int32(&u.V).XdrMarshal(x, x.Sprintf("%sv", name))
	switch u.V {
	case 0:
		x.Marshal(x.Sprintf("%sv0", name), XDR_LedgerCloseMetaV0(u.V0()))
		return
	}
	XdrPanic("invalid V (%v) in LedgerCloseMeta", u.V)
}
```

The marshaller is going to get applied on the union tag field and it will set `u.V` to `randomMarshaller.tag` which is what we want, so that's good.

However, the same marshaller will then be applied on the union arm. So, any uint32 or enum fields contained within the arm will be fixed to `randomMarshaller.tag` which we don't want to happen

Reply:
Ah yeah, I was thinking in the non-union ones you would clear the tag. But I think you're right. 🤔 Ok. gefn.